packages/libs/eda/src/lib/core/components/visualizations/implementations/BipartiteNetworkVisualization.tsx

```diff
@@ -58,47 +65,163 @@ export const BipartiteNetworkConfig = t.partial({
-  // So we'll make it generalizable by mapping the values of the links.color prop to the palette.
+  // Color palettes live here in the frontend, but the backend decides how to color links (ex. by sign of correlation, or avg degree of parent nodes).
+  // So we'll make assigning colors generalizable by mapping the values of the links.color prop to the palette. As we add
+  // different ways to color links in the future, we can adapt our checks and error messaging.
   const uniqueLinkColors = uniq(
-    data.links.map((link) => link.color ?? DEFAULT_LINK_COLOR_DATA)
+    data.value?.bipartitenetwork.data.links.map(
+      (link) => link.color?.toString() ?? DEFAULT_LINK_COLOR_DATA
+    )
   );
+  if (uniqueLinkColors.length > twoColorPalette.length) {
+    throw new Error(
+      `Found ${uniqueLinkColors.length} link colors but expected only {twoColorPalette.length}.`
+    );
+  }
+  // The link color sent from the backend should be either '-1' or '1', but we'll allow any two unique values. Assigning the domain
+  // in the following way preserves "1" getting mapped to the second color in the palette, even if it's the only
+  // unique value in uniqueLinkColors.
+  const linkColorScaleDomain = uniqueLinkColors.every((val) =>
+    ['-1', '1'].includes(val)
+  )
+    ? ['-1', '1']
+    : uniqueLinkColors;
   const linkColorScale = scaleOrdinal<string>()
-    .domain(uniqueLinkColors)
-    .range(twoColorPalette); // the output palette may change if this visualization is reused in other contexts.
-  const cleanedData: BipartiteNetworkData = {
-    ...data,
-    links: data.links.map((link) => {
-      return {
-        ...link,
-        color: linkColorScale(link.color ?? DEFAULT_LINK_COLOR_DATA),
-      };
+    .domain(linkColorScaleDomain)
+    .range(twoColorPalette); // the output palette may change if this visualization is reused in other contexts (ex. not a correlation app).
+
+  // Clean and finalize data format. Specifically, assign link colors, add display labels
+  const cleanedData = useMemo(() => {
+    if (!data.value) return undefined;
+
+    // Find display labels
+    const nodesWithLabels = data.value.bipartitenetwork.data.nodes.map(
+      (node) => {
+        // node.id is the entityId.variableId
+        const displayLabel = fixVarIdLabel(
+          node.id.split('.')[1],
+          node.id.split('.')[0],
+          entities
+        );
+
+        return {
+          id: node.id,
+          label: displayLabel,
+        };
+      }
+    );
+    return {
+      ...data.value.bipartitenetwork.data,
+      nodes: nodesWithLabels,
+      links: data.value.bipartitenetwork.data.links.map((link) => {
+        return {
+          source: link.source,
+          target: link.target,
+          strokeWidth: Number(link.strokeWidth),
+          color: link.color ? linkColorScale(link.color.toString()) : '#000000',
+        };
+      }),
+    };
+  }, [data, entities, linkColorScale]);
```

Comment:
I think `linkColorScale` is causing the infinite loop with the thumbnail and patch request. All of the logic for `linkColorScale` is outside of the `useMemo` and new renders aren't guaranteed to be referentially equal. We can either move this logic inside of this `useMemo` or upset ESLint by removing `linkColorScale` from the dependency array.

Also, I suggest we specify `data.value` instead of just `data` in this array while we're at it.